EIPS/eip-2938.md

```diff
@@ -0,0 +1,265 @@
+5. Users being able to pay for transaction fees in a token other than ETH (eg. by converting that token into the ETH needed for fees inside the transaction in real-time)
+
+Most of the above use cases are currently possible using intermediaries, most notably the [Gas Station Network](https://www.opengsn.org/) and application-specific alternatives. These implementations are (i) technically inefficient, due to the extra 21000 gas to pay for the relayer, (ii) economically inefficient, as relayers need to make a profit on top of the gas fees that they pay. Additionally, use of intermediary protocols means that these applications cannot simply rely on base ethereum infrastructure and need to rely on extra protocols that have smaller userbases and higher risk of no longer being available at some future date.
+
+Out of the five use cases above, single-tenant AA approximately supports (1) and (2), and multi-tenant AA approximately supports (3) and (4). We discuss the differences between the two tiers in the specification and rationale sections below.
+
+## Specification
+
+### Single Tenant
+
+After `FORK_BLOCK`, the following changes will be recognized by the protocol.
+
+#### Constants
+
+| Constant | Value |
+| - | - |
+| **`AA_ENTRY_POINT`** | `0xffffffffffffffffffffffffffffffffffffffff` |
+| **`AA_TX_TYPE`** | `2` |
+| **`FORK_BLOCK`** | TBD |
+| **`AA_BASE_GAS_COST`** | 15000 |
+
+#### New Transaction Type
+
+A new [EIP-2718](./eip-2718.md) transaction with type `AA_TX_TYPE` is introduced. Transactions of this type are referred to as "AA transactions". Their payload should be interpreted as `rlp([nonce, target, data])`.
+
+The base gas cost of this transaction is set to `AA_BASE_GAS_COST` instead of 21000 to reflect the lack of "intrinsic" ECDSA and signature.
+
+Nonces are processed analogously to existing transactions (check `tx.nonce == tx.target.nonce`, transaction is invalid if this fails, otherwise proceed and immediately set `tx.nonce += 1`).
+
+Note that this transaction type has no intrinsic gas limit; when beginning execution, the gas limit is simply set to the remaining gas in the block (ie. `block.gas_limit` minus gas spent on previous transactions), and the `PAYGAS` opcode (see below) can adjust the gas limit downwards.
+
+#### Transaction-wide global variables
+
+Introduce some new transaction-wide global variables. These variables work similarly (in particular, have similar reversion logic) to the SSTORE refunds counter.
+
+| Variable | Type | Initial value |
+| - | - | - |
+| `globals.transaction_fee_paid` | `bool` | `False if type(tx) == AA_TX_TYPE else True` |
+| `globals.gas_price` | `int` | `0 if type(tx) == AA_TX_TYPE else tx.gas_price` |
+| `globals.gas_limit` | `int` | `0 if type(tx) == AA_TX_TYPE else tx.gas_limit` |
+
+#### `NONCE (0xAB)` Opcode
+
+A new opcode `NONCE (0xAB)` is introduced, with gas cost `G_base`, which pushes the `tx.nonce` onto the stack.
```

Comment:
```suggestion
A new opcode `NONCE (0xAB)` is introduced, with gas cost `G_base`, which pushes the `account.nonce` onto the stack.
```

If `tx.nonce` would be used, I would suggest to name this `TXNONCE`. Luckily however `tx.nonce == account.nonce`, so changing the description is enough.